Kubernetes Executor Documentation Updates (#1811)
astronomer/docs: astro/configure-deployment-resources.md

  <meta name="og:description" content="Modify the resource settings of a Deployment to make sure that your tasks have the CPU and memory required to complete successfully." />
</head>

> After you create an Astro Deployment, you can modify its settings to meet the unique requirements of your organization. Through the Cloud UI and Astro CLI, you can:

Reviewer: I still think we should say "using" instead of "in", especially for the Astro CLI where you're not really "in" it

- You have a high number of short-running tasks and want to ensure low latency between tasks.
- You don't often experience conflicts between Python or OS-level packages and don't require dependency isolation.

If your team regularly experiences dependency conflicts or finds that resource intensive tasks consume the resources of other tasks and cause them to fail, Astronomer recommends implementing worker queues or moving to the Kubernetes executor.

See [Manage the Celery executor](configure-executors.md#manage-the-celery-executor) to learn more about how to configure the Celery executor.

### Kubernetes executor

The Kubernetes executor runs each task in an individual Kubernetes Pod instead of in shared Celery workers. For each task that needs to run, the executor calls the Kubernetes API to dynamically launch a Pod for the task. You can specify the configuration of the task and Pod, including CPU and memory, in a `pod_override` file. When the task completes, the Pod terminates. On Astro, the Kubernetes infrastructure required to run the Kubernetes executor is built into every Deployment and is managed by Astronomer.

> The Kubernetes executor is a good fit for teams that need fine-grained configurations for each of their tasks. Specifically, the Kubernetes executor is a good fit for your team if:

Reviewer: ```suggestion
The Kubernetes executor is a good fit for teams that want fine-grained control over the execution environment for each of their tasks. Specifically, the Kubernetes executor is a good fit for your team if:
```

> The Kubernetes executor runs each task in an individual Kubernetes Pod instead of in shared Celery workers. For each task that needs to run, the executor calls the Kubernetes API to dynamically launch a Pod for the task. You can specify the configuration of the task and Pod, including CPU and memory, in a `pod_override` file. When the task completes, the Pod terminates. On Astro, the Kubernetes infrastructure required to run the Kubernetes executor is built into every Deployment and is managed by Astronomer.

Reviewer 2: >You can specify the configuration of the task and Pod, including CPU and memory, in a `pod_override` file.

You can specify the configuration of the task and Pod, including CPU and memory, as part of your DAG definition using the official [Kubernetes Python Client](https://github.com/kubernetes-client/python) and `pod_override` arg.

Reviewer 2: (it's not a separate file - that was the real comment)

- You have long-running tasks that require more than 24 hours to execute. The Kubernetes executor ensures that tasks longer than 24 hours are not interrupted when you deploys code.
- You experience a high number of dependency conflicts between tasks and could benefit from task isolation. For example, one task in your Deployment requires a different version of pandas than another task.
- You have a strong understanding of the CPU and memory that your tasks require and would benefit from being able to allocate and optimize infrastructure resources at the task level.
- You have had issues running certain tasks reliably with the Celery executor.
  
The primary limitation with the Kubernetes executor is that each task takes up to 1 minute to start running once scheduled. If you're running short-running tasks and cannot tolerate high latency, Astronomer recommends the Celery executor. To learn more, see [Kubernetes Executor](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/executor/kubernetes.html) in Airflow documentation or [Manage the Kubernetes executor on Astro](executors.md#manage-the-kubernetes-executor).

> If you want to run only specific tasks in a Kubernetes Pod, consider using the [KubernetesPodOperator](kubernetespodoperator.md).

Reviewer 2: > If you want to run only specific tasks in a Kubernetes Pod

If you want to run arbitrary non-airflow images in a Kubernetes Pod controlled by Airflow, consider using the [KubernetesPodOperator](kubernetespodoperator.md).

### Update the Deployment executor

1. In the Cloud UI, select a Workspace, click **Deployments**, and then select a Deployment.
2. Click the **Details** tab.
3. Click **Edit Details**. 
4. Select **Celery** or **Kubernetes** in the **Executor** list. If you're moving from the Celery to the Kubernetes executor, all existing worker queues are deleted. Running tasks stop gracefully and all new tasks start with the selected executor.
5. Click **Update**.

## Scheduler resources

> The [Airflow scheduler](https://airflow.apache.org/docs/apache-airflow/stable/concepts/scheduler.html) is responsible for monitoring task execution and triggering downstream tasks when the dependencies are met. To ensure that your tasks have the CPU and memory required to complete successfully on Astro, you can set:

Reviewer: ```suggestion
The [Airflow scheduler](https://airflow.apache.org/docs/apache-airflow/stable/concepts/scheduler.html) is responsible for monitoring task execution and triggering downstream tasks when the dependencies are met. 

Scheduler resources must be set for each Deployment and are managed separately from cluster-level infrastructure. Any additional components that Astro requires, including PgBouncer, KEDA, and the triggerer, are managed by Astronomer.

To ensure that your tasks have the CPU and memory required to complete successfully on Astro, you can set:
```

- **Scheduler Resources**: Determine the total CPU and memory allocated to each scheduler in your Deployment, defined as Astronomer Units (AU). One AU is equivalent to 0.1 CPU and 0.375 GiB of memory. The default scheduler size is 5 AU, or .5 CPU and 1.88 GiB memory. The number of schedulers running in your Deployment is determined by **Scheduler Count**, but all schedulers are created with the same CPU and memory allocations.

- **Scheduler Count**: Move the slider to select the number of schedulers for the Deployment. Each scheduler is provisioned with the AU you specified in the **Scheduler Resources** field. For example, if you set scheduler resources to 10 AU and **Scheduler Count** to 2, your Deployment will run with 2 Airflow schedulers using 10 AU each. For high availability, Astronomer recommends selecting a minimum of two schedulers. 

> Scheduler resources must be set for each Deployment and are managed separately from cluster-level infrastructure. Any additional components that Astro requires, including PgBouncer, KEDA, and the triggerer, are managed by Astronomer.

Reviewer: Sure, also I think we can remove that sentence about everything else that Astronomer manages

- Allocate resources for the Airflow scheduler and workers.
- Update a Deployment name and description.
- Add or delete a Deployment alert email.
- Change the Deployment executor.
- Transfer a Deployment to another Workspace in your Organization.
- Delete a Deployment.

For advanced Deployment resource configurations, see [Manage Deployment executors](configure-executors.md) and [Configure worker queues](configure-worker-queues.md).

:::cli

This document focuses on configuring Deployments through the Cloud UI. For steps and best practices for configuring Deployments as code using the Astro CLI, see [Manage Deployments as code](manage-deployments-as-code.md).

:::

## Choose an executor

A Deployment's executor type determines which worker resources run your scheduled tasks. Every Deployment requires an executor and you can update the executor at any time. Astro supports two executors, both of which are available in the Apache Airflow open source project:

- Celery executor
- Kubernetes executor

Use the following topics to understand the benefits and limitations of each executor to determine which one is right for your Deployment. 

### Celery executor

> The Celery executor is the default for all new Deployments. It utilizes a pool of workers and communicates with them to delegate tasks. Astronomer uses [Worker autoscaling logic](configure-executors.md#worker-autoscaling-logic) to determine how many workers run on each worker queue on your Deployment at a given time. 
> 
> The Celery executor manages a pool of worker Pods that are each allocated a certain amount of CPU and memory. A single worker Pod can run multiple tasks at a time. Astronomer uses [Worker autoscaling logic](configure-executors.md#worker-autoscaling-logic) to determine how many workers run on each worker queue on your Deployment at a given time.

Reviewer: ```suggestion
The Celery executor is the default for all new Deployments. It utilizes a pool of workers and communicates with them to delegate tasks. Astronomer uses [Worker autoscaling logic](configure-executors.md#worker-autoscaling-logic) to determine how many workers run on each worker queue on your Deployment at a given time.
```

- Allocate resources for the Airflow scheduler and workers.
- Update a Deployment name and description.
- Add or delete a Deployment alert email.
- Change the Deployment executor.
- Transfer a Deployment to another Workspace in your Organization.
- Delete a Deployment.

For advanced Deployment resource configurations, see [Manage Deployment executors](manage-astro-executors.md) and [Configure worker queues](configure-worker-queues.md).

## Choose an executor

The executor type you select determines which worker resources run your scheduled tasks. A single executor is assigned to each Deployment and you can update the executor assignment at any time. 

### Celery executor

> The Celery executor works with a pool of workers and communicates with them to delegate tasks and it's a good option for most uses cases. Astronomer uses worker autoscaling logic to determine how many workers run on each worker queue on your Deployment at a given time. See [Worker autoscaling logic](manage-astro-executors.md#worker-autoscaling-logic). 

Reviewer: ```suggestion
The Celery executor works with a pool of workers and communicates with them to delegate tasks and it's a good option for most use cases. Astronomer uses worker autoscaling logic to determine how many workers run on each worker queue on your Deployment at a given time. See [Worker autoscaling logic](manage-astro-executors.md#worker-autoscaling-logic). 
```

#### Benefits

> - Worker queues let you assign tasks to different worker types. See [Configure worker queues](configure-worker-queues.md)

Reviewer: ```suggestion
- Support multiple worker queues, which lets you assign tasks to different worker node types. See [Configure worker queues](configure-worker-queues.md)
```

- **Scheduler Resources**: Select the number of Astronomer Units (AU) assigned to the schedulers. Alternatively, move the CPU and memory slider to use specific CPU and memory values to automatically define the scheduler AU assignment. An AU is a unit of CPU and memory allocated to each scheduler in a Deployment. One AU is equivalent to 0.1 CPU and 0.375 GiB of memory. Assigning five AUs to a scheduler is equivalent to 0.5 CPUs and 1.88 GiB of memory. 

    If you experience delays in task execution, which you can track with the Gantt Chart view of the Airflow UI, Astronomer recommends increasing the allocated AU value. The default resource allocation is ten AU. 

- **Scheduler Count**: Move the slider to select the number of schedulers for the Deployment. Each scheduler is provisioned with the AU you specified in the **Scheduler Resources** field. For example, if you set scheduler resources to ten AU and **Scheduler Count** to two, your Deployment will run with two Airflow schedulers using ten AU each. For high availability, Astronomer recommends selecting a minimum of two schedulers. 

> - The [Airflow scheduler](https://airflow.apache.org/docs/apache-airflow/stable/concepts/scheduler.html), which is responsible for monitoring task execution and triggering downstream tasks when the dependencies are met.

Reviewer: ```suggestion
```

See [Manage the Celery executor](manage-astro-executors.md#manage-the-celery-executor) for more information about configuring and running the Kubernetes executor on a Deployment. 

#### Benefits

- Supports multiple worker queues, which lets you assign tasks to different worker node types. See [Configure worker queues](configure-worker-queues.md)
- Scales workers with higher demand.
- Provides a grace period for worker termination.

#### Limitations

- Tasks can run on the same worker, meaning that it's possible for a task use resources intended for another task.

### Kubernetes executor

The Kubernetes executor runs all tasks in individual Kubernetes Pods. You can specify the configuration for these Pods using `pod_override` files which you can then apply to specific tasks. This executor is recommended when you need to control resource optimization, isolate your workloads, run tasks for extended periods, or have extended periods without without task runs.

> See [Manage the Kubernetes executor](manage-astro-executors.md#manage-the-kubernetes-executor) for more information about configuring and running the Kubernetes executor on a Deployment. 

Reviewer: Love all of this but not sure about "for teams running Airflow at scale." Isn't that Astro in general, and thus all of our customers? What about "Good fit for customers that need fine-grained configurations for individual tasks"

- **Scheduler Count**: Move the slider to select the number of schedulers for the Deployment. Each scheduler is provisioned with the AU you specified in the **Scheduler Resources** field. For example, if you set scheduler resources to ten AU and **Scheduler Count** to two, your Deployment will run with two Airflow schedulers using ten AU each. For high availability, Astronomer recommends selecting a minimum of two schedulers. 

Scheduler resources must be set for each Deployment and are managed separately from cluster-level infrastructure. Any additional components that Astro requires, including PgBouncer, KEDA, and the triggerer, are managed by Astronomer.

:::cli

If you prefer, you can set Deployment resources using the Astro CLI and a local Deployment configuration file. See [Deployments as Code](manage-deployments-as-code.md).

:::

### Edit scheduler settings 

1. In the Cloud UI, select a Workspace, click **Deployments**, and then select a Deployment.
2. Click the **Details** tab.
3. Click **Edit Details**. 
4. Edit the scheduler resource settings. See [Scheduler resources](#scheduler-resources).

> 5. Click **Update**.

Reviewer: Fixed this so that it's parallel to our executor section.

- Allocate resources for the Airflow scheduler and workers.
- Update a Deployment name and description.
- Add or delete a Deployment alert email.
- Change the Deployment executor.
- Transfer a Deployment to another Workspace in your Organization.

> - Delete a Deployment.

Reviewer: ```suggestion
- Delete a Deployment.

For information on more advanced Deployment resource configurations, see [Manage Deployment executors](manage-astro-executors.md) and [Configure worker queues](configure-worker-queues.md).
```

- Running tasks are not interrupted when a deploy is pushed.

On Astro, the Kubernetes infrastructure required to run the Kubernetes executor is built into every cluster in the data plane and is managed by Astronomer.

#### Limitations

- Tasks take longer to start and this causes task latency.
- PersistentVolumes (PVs) are not supported on Pods launched in an Astro cluster.
- The `pod_template_file` argument is not supported on Pods launched in an Astro cluster. If you use the `pod_template_file` argument, the DAG is rejected and a broken DAG error message appears in teh Airflow UI. Astronomer recommends using `python-kubernetes-sdk`. See [Astro Python SDK ReadTheDocs](https://astro-sdk-python.readthedocs.io/en/stable/).

### Update the Deployment executor

1. In the Cloud UI, select a Workspace, click **Deployments**, and then select a Deployment.
2. Click the **Details** tab.
3. Click **Edit Details**. 
4. Select **Celery** or **Kubernetes** in the **Executor** list. If you're moving from the Celery to the Kubernetes executor, all existing worker queues are deleted. Running tasks stop gracefully and all new tasks start with the selected executor.
5. Click **Update**.

> ## Set Deployment resources

Author: @jwitz I made an attempt to modify the structure as you requested. Please review and make any adjustments you think are necessary.

### Celery executor

The Celery executor works with a pool of workers and communicates with them to delegate tasks and it's a good option for most use cases. Astronomer uses worker autoscaling logic to determine how many workers run on each worker queue on your Deployment at a given time. See [Worker autoscaling logic](manage-astro-executors.md#worker-autoscaling-logic). 

#### Benefits

> - Support multiple worker queues, which lets you assign tasks to different worker node types. See [Configure worker queues](configure-worker-queues.md)

Reviewer: ```suggestion
- Supports multiple worker queues, which lets you assign tasks to different worker node types. See [Configure worker queues](configure-worker-queues.md)
```

### Scheduler resources

> The [Airflow scheduler](https://airflow.apache.org/docs/apache-airflow/stable/concepts/scheduler.html) is responsible for monitoring task execution and triggering downstream tasks when the dependencies are met. The executor type you select determines what scheduler resource settings can be adjusted. You can select a Celery or a Kubernetes executor. The executor determines which worker resources run your scheduled tasks.

Author: @cmarteepants the text has been relocated.

You can adjust the following scheduler resource settings:

- **Default Max Tasks Per Worker**: Celery executor only. Move the slider to select the number of tasks that a worker can process simultaneously. 

> - **Default Worker Count**: Celery executor only. Move the slider to select the number of can run in parallel in the worker queue.

Author: @cmarteepants I made the changes you requested (I think). It wasn't absolutely clear what you meant by "...move the worker queues section under the scheduler resources section". I assumed you wanted it to be a child of S_cheduler resources_.

- **Scheduler Resources**: Select the number of Astronomer Units (AU) assigned to the schedulers. Alternatively, move the CPU and memory slider to use specific CPU and memory values to automatically define the scheduler AU assignment. An AU is a unit of CPU and memory allocated to each scheduler in a Deployment. One AU is equivalent to 0.1 CPU and 0.375 GiB of memory. Assigning five AUs to a scheduler is equivalent to 0.5 CPUs and 1.88 GiB of memory. 
- **Scheduler Count**: Move the slider to select the number of schedulers for the Deployment. Each scheduler is provisioned with the AU you specified in the **Scheduler Resources** field. For example, if you set scheduler resources to ten AU and **Scheduler Count** to two, your Deployment will run with two Airflow schedulers using ten AU each. For high availability, Astronomer recommends selecting a minimum of two schedulers. 

If you experience delays in task execution, which you can track with the Gantt Chart view of the Airflow UI, Astronomer recommends increasing the allocated AU value. The default resource allocation is ten AU. 

> #### Edit scheduler settings 

Author: @cmarteepants I implemented the change as requested.

1. In the Cloud UI, select a Workspace, click **Deployments**, and then select a Deployment.
2. Click the **Details** tab.
3. Click **Edit Details**. 

> 4. Optional. Select an executor in the **Executor** list. See [Update the Deployment executor](#update-the-deployment-executor).

Reviewer: Per discussion a few days ago, I would cut any steps that are unrelated to scheduler settings.

Author: @jwitz the text has been removed.

5. Edit the scheduler resource settings. See [Scheduler resources](#scheduler-resources).
6. Click **Update**.

>     The Airflow components of your Deployment automatically restart to apply the updated resource allocations. This action is equivalent to deploying code to your Deployment and does not impact running tasks that have 24 hours to complete before running workers are terminated. See [What happens during a code deploy](deploy-code.md#what-happens-during-a-code-deploy).

Author: @cmarteepants I have modified the text.

### Update the Deployment executor

> The executor determines which worker resources run your scheduled tasks. The Celery executor runs multiple tasks on a single pod and is a good option for most uses cases. The Kubernetes executor runs individual tasks in an isolated Kubernetes pod and is a good option when a few seconds of latency during start up isn't a concern. For more information about the Celery and Kubernetes executors, see [Airflow Executors](https://docs.astronomer.io/learn/airflow-executors-explained#kubernetes-executor).

Reviewer 2: > good option when a few seconds of latency during start up isn't a concern

good option when task isolation and reliability are the priority and startup latency isn't a concern

Author: @cmarteepants I added the changes

> The executor determines which worker resources run your scheduled tasks. The Celery executor runs multiple tasks on a single pod and is a good option for most uses cases. The Kubernetes executor runs individual tasks in an isolated Kubernetes pod and is a good option when a few seconds of latency during start up isn't a concern. For more information about the Celery and Kubernetes executors, see [Airflow Executors](https://docs.astronomer.io/learn/airflow-executors-explained#kubernetes-executor).

Reviewer: ```suggestion
The executor determines which worker resources run your scheduled tasks. The Celery executor runs tasks using a pool of worker nodes is a good option for most uses cases. The Kubernetes executor runs individual tasks in an isolated Kubernetes Pod and is a good option when a few seconds of latency during start up isn't a concern. For more information about how to configure your Deployment to use the Celery and Kubernetes executors, see [Manage executors](manage-astro-executors.md).
```

1. In the Cloud UI, select a Workspace, click **Deployments**, and then select a Deployment.
2. Click the **Details** tab.
3. Click **Edit Details**. 
4. Select **Celery** or **Kubernetes** in the **Executor** list. If you're moving from the Celery to the Kubernetes executor, all existing worker queues are deleted. Running tasks stop gracefully and all new tasks start with the selected executor.

> 5. Optional. Edit the scheduler resource settings. See [Scheduler resources](#scheduler-resources).

Reviewer: Cut this

You can adjust the following scheduler resource settings:

- **Scheduler Resources**: Select the number of Astronomer Units (AU) assigned to the schedulers. Alternatively, move the CPU and memory slider to use specific CPU and memory values to automatically define the scheduler AU assignment. An AU is a unit of CPU and memory allocated to each scheduler in a Deployment. One AU is equivalent to 0.1 CPU and 0.375 GiB of memory. Assigning five AUs to a scheduler is equivalent to 0.5 CPUs and 1.88 GiB of memory. 
- **Scheduler Count**: Move the slider to select the number of schedulers for the Deployment. Each scheduler is provisioned with the AU you specified in the **Scheduler Resources** field. For example, if you set scheduler resources to ten AU and **Scheduler Count** to two, your Deployment will run with two Airflow schedulers using ten AU each. For high availability, Astronomer recommends selecting a minimum of two schedulers. 

If you experience delays in task execution, which you can track with the Gantt Chart view of the Airflow UI, Astronomer recommends increasing the allocated AU value. The default resource allocation is ten AU. 

> #### Worker queues 

Author: @jwitz sorry, it's unclear what specific edit you're requesting here. Please implement the change you think is necessary.

> #### Worker queues 

Reviewer: See my most recent commit

Worker queues are a set of configurations that apply to a group of workers in your Deployment. Each Deployment includes a `default` worker queue for running tasks, but you can configure additional worker queues to define CPU and memory limits for your tasks.

You can adjust the following worker queue settings:

- **Default Max Tasks Per Worker**: Celery executor only. Move the slider to select the number of tasks that a worker can process simultaneously. 
- **Default Worker Count**: Celery executor only. Move the slider to select the number of can run in parallel in the worker queue.

See [Configure worker queues](configure-worker-queues.md).

> #### Edit Deployment settings 

Reviewer: ```suggestion
### Edit scheduler settings 
```
This topic seems to only be about schedulers now

1. In the Cloud UI, select a Workspace, click **Deployments**, and then select a Deployment.
2. Click the **Details** tab.
3. Click **Edit Details**. 
4. Edit the scheduler resource settings. See [Scheduler resources](#scheduler-resources).
5. Click **Update**.

>     The Airflow components of your Deployment automatically restart to apply the updated resource allocations. This action is equivalent to deploying code to your Deployment and does not impact running tasks that have 24 hours to complete (Celery executor only) before running workers are terminated. See [What happens during a code deploy](deploy-code.md#what-happens-during-a-code-deploy).

Reviewer: ```suggestion
    The Airflow components of your Deployment automatically restart to apply the updated resource allocations. This action is equivalent to deploying code and will trigger a rebuild of your Deployment image. If you're using the Celery executor, currently running tasks have 24 hours to complete before their running workers are terminated. See [What happens during a code deploy](deploy-code.md#what-happens-during-a-code-deploy).
```
